new package: callflow

bhatiaharsh opened the pull request:
new package: `callflow`.
updates to: `node-js` (add a variant that allows installing with `npm`)
updates to: `py-greenlet` and `npm` (new versions and checksum)

bhatiaharsh committed:
new package: callflow

adamjstewart commented on var/spack/repos/builtin/packages/callflow/package.py:
```diff
@@ -0,0 +1,40 @@
+    depends_on('py-matplotlib',     type=('build', 'run'))
+    depends_on('py-networkx',       type=('build', 'run'))
+    depends_on('py-statsmodels',    type=('build', 'run'))
+    depends_on('py-scikit-learn',   type=('build', 'run'))
```
I don't see this dep listed in `setup.py`

bhatiaharsh replied:
`callflow` is still wip. i will fix these as I go through verifying all of this.

bhatiaharsh replied:
updated package.py and setup.py are consistent

adamjstewart commented on var/spack/repos/builtin/packages/callflow/package.py:
```diff
@@ -0,0 +1,40 @@
+    depends_on('py-jsonschema',     type=('build', 'run'))
+
+    depends_on('py-matplotlib',     type=('build', 'run'))
+    depends_on('py-networkx',       type=('build', 'run'))
```
```suggestion
    depends_on('py-networkx@2.2',   type=('build', 'run'))
```

bhatiaharsh replied:
updated package.py and setup.py are consistent

adamjstewart commented on var/spack/repos/builtin/packages/callflow/package.py:
```diff
@@ -0,0 +1,40 @@
+    depends_on('py-colorlog',       type=('build', 'run'))
+    depends_on('py-jsonschema',     type=('build', 'run'))
+
+    depends_on('py-matplotlib',     type=('build', 'run'))
```
I don't see this dep listed in `setup.py`

bhatiaharsh replied:
updated package.py and setup.py are consistent

adamjstewart commented on var/spack/repos/builtin/packages/callflow/package.py:
```diff
@@ -0,0 +1,40 @@
+    depends_on('py-pandas',         type=('build', 'run'))
+    depends_on('py-hatchet',        type=('build', 'run'))
+
+    depends_on('py-pydot',          type=('build', 'run'))
+    depends_on('py-pyyaml',         type=('build', 'run'))
+    depends_on('py-colorlog',       type=('build', 'run'))
+    depends_on('py-jsonschema',     type=('build', 'run'))
```
I don't see any of these deps listed in `setup.py`

bhatiaharsh replied:
updated package.py and setup.py are consistent

adamjstewart commented on var/spack/repos/builtin/packages/node-js/package.py:
```diff
@@ -36,6 +36,7 @@ class NodeJs(Package):
     variant('icu4c', default=False, description='Build with support for all locales instead of just English')
     variant('openssl', default=True,  description='Build with Spacks OpenSSL instead of the bundled version')
     variant('zlib', default=True,  description='Build with Spacks zlib instead of the bundled version')
+    variant('without-npm', default=True, description='Build without npm dependency (to be installed explicitly)')
```
Can you explain why this variant is necessary? Why can't a package depend on both `node-js` and `npm`?

bhatiaharsh replied:
Yes, in almost all cases, someone would want to use both `node-js` and `npm`. In the prev package, `--without-npm` was being passed to configure step, with the following comment.
```
# Note: npm is updated more regularly than node.js, so we build
# the package instead of using the bundled version
```

However, i find this wrong, particularly because `npm` failed for me due to some very deep dependency incompatibility. Here, I propose putting `--without-npm` option as a variant, so people can choose. e.g., it worked with me to install `node-js` with `npm`.

adamjstewart commented on var/spack/repos/builtin/packages/py-greenlet/package.py:
```diff
@@ -10,6 +10,7 @@ class PyGreenlet(PythonPackage):
 
     homepage = "https://github.com/python-greenlet/greenlet"
-    url      = "https://pypi.io/packages/source/g/greenlet/greenlet-0.4.13.tar.gz"
+    url      = "https://github.com/python-greenlet/greenlet/archive/0.4.17.tar.gz"
```
Can you explain why this URL change is necessary? It changes the checksum of the previous version. Also see https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#url

bhatiaharsh replied:
good point. looks like this was a bad call. i will revert this.

bhatiaharsh replied:
fixed this

adamjstewart reviewed (changes requested):
It might be easiest to split things up in separate PRs for each package

bhatiaharsh commented:
i think this should be good to go, if you dont mind pulling all the changes in from a single PR.

adamjstewart commented on var/spack/repos/builtin/packages/callflow/package.py:
```diff
@@ -0,0 +1,72 @@
+    maintainers = ["bhatiaharsh", "jarusified"]
+
+    version('1.1.0', sha256='f8b875eb62fbac04b117e3c23fccff99d768158226a9b7fa222a2b2a6acafa44')
+    version('v1.1.1', tag='v1.1.1')
```
Why is one version downloaded from a checksum and the other from a git tag? Why is one version prefixed with `v` and the other isn't? You should also sort versions from newest to oldest.

adamjstewart commented on var/spack/repos/builtin/packages/callflow/package.py:
```diff
@@ -0,0 +1,72 @@
+
+    homepage = "https://github.com/LLNL/CallFlow"
+    url      = "https://github.com/LLNL/CallFlow/archive/v1.1.0.tar.gz"
+    git      = 'git@github.com:LLNL/CallFlow.git'
```
```suggestion
    git      = 'https://github.com/LLNL/CallFlow.git'
```
This seems to be more reliable if you don't have SSH keys setup.

adamjstewart commented on var/spack/repos/builtin/packages/callflow/package.py:
```diff
@@ -0,0 +1,72 @@
+    @run_after('build')
+    def build_client(self):
+
+        src = os.path.join(os.getcwd(), 'app')
+        with working_dir(src):
+            os.system('npm install')
+            os.system('npm run build')
```
```suggestion
        with working_dir('app'):
            npm = which('npm')
            npm('install')
            npm('run', 'build')
```

adamjstewart commented on var/spack/repos/builtin/packages/callflow/package.py:
```diff
@@ -0,0 +1,72 @@
+        fname = os.path.join(self.spec.prefix.bin, 'callflow_app')
+        with open(fname, 'w') as f:
+            f.write('#!/bin/sh\n')
+            f.write('cd {}\n'.format(dst))
```
```suggestion
            f.write('cd {0}\n'.format(dst))
```
Needed for Python 2.6 support

adamjstewart reviewed (commented):
I'm still pretty hesitant about the node-js/npm change. I would much rather get the npm package working than to duplicate it. For Python, we don't install pip, we provide it as a separate package.

bhatiaharsh commented:
> I'm still pretty hesitant about the node-js/npm change. I would much rather get the npm package working than to duplicate it. For Python, we don't install pip, we provide it as a separate package.


hardcoding `~without-npm` in `node-js` package is a handicap and frankly a silly idea. `node` and `npm` are released after consistency checks. why would we not want to leverage that?

I have simply added more flexibility by moving this option to a variant, since that is key here. isnt spack all about flexibility?

the user can still choose to install `npm` separately as you want. but if someone runs into issues, why should they have to go about a wild chase trying to fix an issue that they shouldn't have to worry about? i am strongly against doing that.

bhatiaharsh committed:
added flask-cors

bhatiaharsh commented:
Hey @adamjstewart. we have updated the way callflow is installed, and this now does not require the changes that you objected to (npm and nodejs). therefore, the previous conversation was not relevant -- i have resolved those.

i have now squashed everything into 2x commits only. can you please approve and merge them? I need to deliver the package this week, and spack would be the easiest way. thank you.

adamjstewart commented on var/spack/repos/builtin/packages/callflow/package.py:
```diff
@@ -0,0 +1,43 @@
+    depends_on('py-networkx',       type=('build', 'run'))
+
+    depends_on('py-ipython',        type=('build', 'run'))
+    depends_on('py-flask-socketio', type=('build', 'run'))
+    depends_on('py-flask-cors',     type=('build', 'run'))
```
I don't see these deps in the `setup.py`, are they required?

bhatiaharsh replied:
yes, i am also working through the package repo to fix all dependencies (this is a PR that will go in shortly). but spack dependencies are correct -- i tested in the morning
